File: docs/sdk/x/v3.0(latest)/IPMExample-objc/IPMExample-objc/UI/ViewController.m
//
//  ViewController.m
//  IPMExample-objc
//
//  Created by Pavel Gorb on 4/18/16.
//  Copyright © 2016 Virgil Security, Inc. All rights reserved.
//

#import "ViewController.h"
#import "IPMChannelClient.h"
#import "IPMSecurityManager.h"
#import "IPMSecureMessage.h"
#import "LoadingViewController.h"
#import "IPMConstants.h"

@import VirgilSDK;
@import XAsync;

static NSString * _Nonnull const kIPMChatCell = @"IPMChatCell";

static NSString * _Nonnull const kIPMMessageText = @"text";
static NSString * _Nonnull const kIPMMessageSender = @"sender";

@interface ViewController () <UITableViewDelegate, UITableViewDataSource, UITextFieldDelegate>

@property (nonatomic, weak) IBOutlet UITableView *tvChat;
@property (nonatomic, weak) IBOutlet UILabel *lTitle;
@property (nonatomic, weak) IBOutlet UILabel *lIdentity;
@property (nonatomic, weak) IBOutlet UITextField *tfMessage;

@property (nonatomic, weak) IBOutlet NSLayoutConstraint *lcBottom;

@property (nonatomic, strong) IPMSecurityManager *ipmSecurity;
@property (nonatomic, strong) IPMChannelClient *ipmClient;

@property (nonatomic, strong) NSMutableArray *messages;
@property (nonatomic, strong) LoadingViewController *loadingController;

@end

@implementation ViewController

@synthesize tvChat = _tvChat;
@synthesize lTitle = _lTitle;
@synthesize lIdentity = _lIdentity;
@synthesize tfMessage = _tfMessage;

@synthesize lcBottom = _lcBottom;

@synthesize ipmSecurity = _ipmSecurity;
@synthesize ipmClient = _ipmClient;
@synthesize messages = _messages;
@synthesize loadingController = _loadingController;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    if (self.messages == nil) {
        self.messages = [[NSMutableArray alloc] init];
    }
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShowNotification:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillHideNotification:) name:UIKeyboardWillHideNotification object:nil];
    
    UIStoryboard *main = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    self.loadingController = [main instantiateViewControllerWithIdentifier:@"LoadingViewController"];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    IPMDataSourceListener listener = ^(IPMSecureMessage * _Nonnull content, NSString * _Nonnull sender) {
        if (content != nil) {
            BOOL ok = [self.ipmSecurity checkSignature:content.signature data:content.message identity:sender];
            if (!ok) {
                NSLog(@"Error verifying sender's signature");
                return;
            }
            
            NSData *plainData = [self.ipmSecurity decryptData:content.message];
            if (plainData.length == 0) {
                NSLog(@"Error decrypting message!");
                return;
            }
            
            NSString *text = [[NSString alloc] initWithData:plainData encoding:NSUTF8StringEncoding];
            NSDictionary *dict = @{ kIPMMessageText: text, kIPMMessageSender: sender };
            [self.messages addObject:dict];
            
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.tvChat reloadData];
            });
        }
    };
    
    void (^IPMChannelSetup)(void) = ^(void) {
        self.ipmClient = [[IPMChannelClient alloc] initWithUserId:self.ipmSecurity.identity];
        NSError *joinError = [self.ipmClient joinChannelWithName:kAppChannelName channelListener:listener];
        if (joinError != nil) {
            NSLog(@"Error joining the channel: %@", [joinError localizedDescription]);
            return;
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            [self dismissViewControllerAnimated:NO completion:nil];
            self.lIdentity.text = self.ipmClient.userId;
        });
    };
    
    UIAlertController *introduce = [UIAlertController alertControllerWithTitle:nil message:NSLocalizedString(@"Enter your email.", @"Enter your email") preferredStyle:UIAlertControllerStyleAlert];
    [introduce addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.keyboardType = UIKeyboardTypeEmailAddress;
        textField.returnKeyType = UIReturnKeyDone;
    }];
    
    UIAlertAction *ok = [UIAlertAction actionWithTitle:NSLocalizedString(@"Ok", @"Ok") style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        UITextField *tf = introduce.textFields[0];
        if (tf.text.length > 0) {
            [self presentViewController:self.loadingController animated:NO completion:nil];
            self.ipmSecurity = [[IPMSecurityManager alloc] initWithIdentity:tf.text];
            [self dismissViewControllerAnimated:NO completion:nil];
            
            NSError *signError = [self.ipmSecurity signin];
            if (signError != nil) {
                if (signError.code == -5555) {
                    /// There is no card found. Need to sign up.
                    NSError *signupError = [self.ipmSecurity signup];
                    if (signupError != nil) {
                        [self dismissViewControllerAnimated:NO completion:nil];
                        NSLog(@"Error: %@", [signupError localizedDescription]);
                        return;
                    }
                    
                    IPMChannelSetup();
                    return;
                }
                
                [self dismissViewControllerAnimated:NO completion:nil];
                NSLog(@"Error: %@", [signError localizedDescription]);
                return;
            }
            
            IPMChannelSetup();
        }
    }];
    [introduce addAction:ok];
    [self presentViewController:introduce animated:YES completion:nil];
}

#pragma mark - Private class logic

- (void)sendMessage:(NSString *)message {
    [self presentViewController:self.loadingController animated:NO completion:nil];
    NSArray *participants = [self.ipmClient.channel getParticipants];
    if (participants.count == 0) {
        [self dismissViewControllerAnimated:NO completion:nil];
        NSLog(@"There is no participants on the channel.");
        return;
    }
    
    NSData *encrypted = [self.ipmSecurity encryptData:[message dataUsingEncoding:NSUTF8StringEncoding allowLossyConversion:NO] identities:participants];
    if (encrypted.length == 0) {
        [self dismissViewControllerAnimated:NO completion:nil];
        NSLog(@"Error encrypting message.");
        return;
    }
    
    NSData *signature = [self.ipmSecurity composeSignatureOnData:encrypted];
    if (signature.length == 0) {
        [self dismissViewControllerAnimated:NO completion:nil];
        NSLog(@"Error composing the signature.");
        return;
    }
    
    IPMSecureMessage *sm = [[IPMSecureMessage alloc] initWithMessage:encrypted signature:signature];
    NSError *sendError = [self.ipmClient.channel sendMessage:sm];
    if (sendError != nil) {
        [self dismissViewControllerAnimated:NO completion:nil];
        NSLog(@"Error sending the message: %@", [sendError localizedDescription]);
        return;
    }
    
    self.tfMessage.text = @"";
    [self dismissViewControllerAnimated:NO completion:nil];
//
//    
//    BOOL ok = [self.ipmSecurity checkSignature:signature data:encrypted identity:self.ipmSecurity.identity];
//    if (!ok) {
//        NSLog(@"<<<<<<<<<<<<<Error verifying sender's signature");
//        return;
//    }
//    NSData *plainData = [self.ipmSecurity decryptData:encrypted];
//    if (plainData.length == 0) {
//        NSLog(@"<<<<<<<<<<<<Error decrypting message!");
//        return;
//    }
//    NSString *text = [[NSString alloc] initWithData:plainData encoding:NSUTF8StringEncoding];
//    if (![text isEqualToString:message]) {
//        NSLog(@">>>>>>>>>>>>>>FUCKED!");
//    }
//    self.tfMessage.text = @"";
//    [self dismissViewControllerAnimated:NO completion:nil];
}

#pragma mark - UITableViewDataSource

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.messages count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kIPMChatCell];
    
    NSDictionary *content = [self.messages[indexPath.row] as:[NSDictionary class]];
    cell.textLabel.text = content[kIPMMessageText];
    cell.detailTextLabel.text = content[kIPMMessageSender];
    return cell;
}

#pragma mark - UITableViewDelegate



#pragma mark - UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField.text.length > 0) {
        [self sendMessage:textField.text];
    }
    [textField resignFirstResponder];
    return YES;
}

#pragma mark - UIKeyboard notifications

- (void)keyboardWillShowNotification:(NSNotification *)notification {
    NSDictionary *userInfo = [notification userInfo];
    
    CGRect endFrame = [([userInfo[UIKeyboardFrameEndUserInfoKey] as:[NSValue class]]) CGRectValue];
    CGRect convertedFrame = [self.view convertRect:endFrame fromView:self.view.window];
    
    self.lcBottom.constant = convertedFrame.size.height;
    [self.view setNeedsLayout];
    [self.view layoutIfNeeded];
}

- (void)keyboardWillHideNotification:(NSNotification *)notification {
    self.lcBottom.constant = 20.0;
    [self.view setNeedsLayout];
    [self.view layoutIfNeeded];
}

@end

File: docs/sdk/x/v3.0(latest)/IPMExample-swift/IPMExample-swift/UI/ViewController.swift
//
//  ViewController.swift
//  IPMExample-swift
//
//  Created by Pavel Gorb on 4/18/16.
//  Copyright © 2016 Virgil Security, Inc. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, UITextFieldDelegate {

    @IBOutlet private var tvChat: UITableView!
    @IBOutlet private var lTitle: UILabel!
    @IBOutlet private var lIdentity: UILabel!
    @IBOutlet private var tfMessage: UITextField!
    @IBOutlet private var lcBottom: NSLayoutConstraint!
    
    private var loadingController: LoadingViewController! = nil
    private var messages = [Dictionary<String, String>]()

    private var ipmSecurity: IPMSecurityManager! = nil
    private var ipmClient: IPMChannelClient! = nil
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        self.loadingController = storyboard.instantiateViewControllerWithIdentifier("LoadingViewController") as! LoadingViewController
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(self.keyboardWillShowNotification(_:)), name: UIKeyboardWillShowNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(self.keyboardWillHideNotification(_:)), name: UIKeyboardWillHideNotification, object: nil)
    }

    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        let listener: IPMDataSourceListener = { secureMessage, sender in
            if !self.ipmSecurity.checkSignature(secureMessage.signature, data: secureMessage.message, identity: sender) {
                print("Error verifying sender's signature.")
                return
            }
            
            if let plainData = self.ipmSecurity.decryptData(secureMessage.message), text = NSString(data: plainData, encoding: NSUTF8StringEncoding) {
                self.messages.append([kIPMMessageText: text as String, kIPMMessageSender: sender])
                
                dispatch_async(dispatch_get_main_queue()) {
                    self.tvChat.reloadData()
                }
                return
            }
            
            print("Error decrypting sender's message.")
        }
        
        let ipmChannelSetup = {
            self.ipmClient = IPMChannelClient(userId: self.ipmSecurity.identity)
            if let error = self.ipmClient.joinChannel(kAppChannelName, listener: listener) {
                dispatch_async(dispatch_get_main_queue()) {
                    self.dismissViewControllerAnimated(false, completion: nil)
                }
                print("Error joining the channel: \(error.localizedDescription)")
                return
            }

            dispatch_async(dispatch_get_main_queue()) {
                self.dismissViewControllerAnimated(false, completion: nil)
                self.lIdentity.text = self.ipmSecurity.identity
            }
        }
        
        let introduction = UIAlertController(title: nil, message: NSLocalizedString("Enter your email", comment: ""), preferredStyle: .Alert)
        introduction.modalPresentationStyle = .OverCurrentContext
        introduction.addTextFieldWithConfigurationHandler { textField in
            textField.keyboardType = .EmailAddress
            textField.returnKeyType = .Done
        }
        let ok = UIAlertAction(title: NSLocalizedString("Ok", comment: ""), style: .Default) { _ in
            let tf = introduction.textFields![0]
            if let text = tf.text {
                self.presentViewController(self.loadingController, animated: false, completion: nil)
                self.ipmSecurity = IPMSecurityManager(identity: text)
                if let signinerror = self.ipmSecurity.signin() {
                    if signinerror.code == -5555 {
                        if let signuperror = self.ipmSecurity.signup() {
                            self.dismissViewControllerAnimated(false, completion: nil)
                            print("Error: \(signuperror.localizedDescription)")
                            return
                        }
                        
                        ipmChannelSetup()
                        return
                    }
                    
                    self.dismissViewControllerAnimated(false, completion: nil)
                    print("Error: \(signinerror.localizedDescription)")
                    return
                }
                
                ipmChannelSetup()
            }
        }
        
        introduction.addAction(ok)
        self.presentViewController(introduction, animated: true, completion: nil)
    }

    func sendMessage(text: String) {
        self.presentViewController(self.loadingController, animated: false, completion: nil)
        
        let participants = self.ipmClient.channel.getParticipants()
        if participants.count > 0 {
            if let plaindata = text.dataUsingEncoding(NSUTF8StringEncoding, allowLossyConversion: false) {
                if let encrypted = self.ipmSecurity.encryptData(plaindata, identities: participants) {
                    if let signature = self.ipmSecurity.composeSignatureOnData(encrypted) {
                        let ipm = IPMSecureMessage(message: encrypted, signature: signature)
                        if let error = self.ipmClient.channel.sendMessage(ipm) {
                            print("Error: \(error.localizedDescription)")
                            self.dismissViewControllerAnimated(false, completion: nil)
                            return
                        }
                        /// success
                        self.tfMessage.text = nil
                        self.dismissViewControllerAnimated(false, completion: nil)
                        return
                    }
                }
            }
        }
        
        print("No participants on the channel.")
        self.dismissViewControllerAnimated(false, completion: nil)
    }
    
    func keyboardWillShowNotification(notification: NSNotification) {
        let userInfo = notification.userInfo!
        
        let keyboardEndFrame = (userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue).CGRectValue()
        let convertedKeyboardEndFrame = view.convertRect(keyboardEndFrame, fromView: view.window)
        
        self.lcBottom.constant = convertedKeyboardEndFrame.size.height
        self.view.setNeedsLayout()
        self.view.layoutIfNeeded()
    }
    
    func keyboardWillHideNotification(notification: NSNotification) {
        self.lcBottom.constant = 20.0
        self.view.setNeedsLayout()
        self.view.layoutIfNeeded()
    }
}

extension ViewController {

    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.messages.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier(kIPMChatCell, forIndexPath: indexPath)
        
        let content = self.messages[indexPath.row]
        cell.textLabel?.text = content[kIPMMessageText]
        cell.detailTextLabel?.text = content[kIPMMessageSender]
        
        return cell
    }
    
}

extension ViewController {
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if let text = textField.text {
            self.sendMessage(text)
        }
        
        textField.resignFirstResponder()
        return true
    }
    
}

